File: src/combat.py
"""Turn-based combat system."""
import random

from src import ui
from src.data_loader import load_data
from src.player import Player
from src.world import World


def calculate_damage(attacker_attack: int, defender_defense: int, speed: int) -> int:
    """Calculate combat damage."""
    base = max(1, attacker_attack - defender_defense * 0.5)
    variance = random.uniform(0.9, 1.1)
    crit_chance = speed * 0.005
    crit = random.random() < crit_chance
    damage = int(base * variance * (2 if crit else 1))
    return max(1, damage), crit


def start_combat(player: Player, npc: dict, world: World):
    """Start a combat encounter. Returns True if player wins."""
    npc_data = npc.get("combat", {})
    npc_hp = npc_data.get("hp", 50)
    npc_max_hp = npc_hp
    npc_mp = npc_data.get("mp", 0)
    npc_attack = npc_data.get("attack", 10)
    npc_defense = npc_data.get("defense", 5)
    npc_speed = npc_data.get("speed", 10)
    npc_name = npc.get("name", "敌人")

    ui.clear()
    ui.print_header("战斗开始")
    print(f"你对 {npc_name} 发起了攻击！\n")

    while player.is_alive() and npc_hp > 0:
        ui.print_line(f"【{player.name}】 HP: {player.hp}/{player.max_hp}  MP: {player.mp}/{player.max_mp}")
        ui.print_line(f"【{npc_name}】   HP: {npc_hp}/{npc_max_hp}\n")

        choice = ui.show_menu("战斗", [
            ("attack", "攻击"),
            ("skill", "使用武功"),
            ("item", "使用物品"),
            ("flee", "逃跑"),
        ])

        if choice == "":
            continue

        if choice == "flee":
            flee_chance = 0.3 + (player.speed - npc_speed) * 0.02
            if random.random() < flee_chance:
                ui.print_line("\n你趁机逃走了！")
                ui.pause()
                return False
            else:
                ui.print_line("\n逃跑失败！")

        elif choice == "attack":
            damage, crit = calculate_damage(player.total_attack(), npc_defense, player.speed)
            npc_hp -= damage
            msg = f"你攻击 {npc_name}，造成 {damage} 点伤害"
            if crit:
                msg += "（暴击！）"
            ui.print_line(msg)

        elif choice == "skill":
            if not player.skills:
                ui.print_line("你还没有学会任何武功。")
                continue
            skill_options = []
            for skill_id, level in player.skills.items():
                skill = world.get_skill(skill_id)
                if skill:
                    skill_options.append((skill_id, f"{skill['name']} Lv.{level}"))
            skill_id = ui.show_menu("选择武功", skill_options)
            if not skill_id:
                continue
            skill = world.get_skill(skill_id)
            damage, crit = calculate_damage(
                player.total_attack() + player.skills[skill_id] * 2,
                npc_defense,
                player.speed
            )
            npc_hp -= damage
            ui.print_line(f"你使出 {skill['name']}，造成 {damage} 点伤害！")

        elif choice == "item":
            if not player.inventory:
                ui.print_line("背包空空如也。")
                continue
            item_options = []
            for item_id, qty in player.inventory.items():
                item = world.get_item(item_id)
                if item and item["type"] == "consumable":
                    item_options.append((item_id, f"{item['name']} x{qty}"))
            if not item_options:
                ui.print_line("没有可使用的物品。")
                continue
            item_id = ui.show_menu("使用物品", item_options)
            if not item_id:
                continue
            item = world.get_item(item_id)
            effect = item.get("effect", {})
            if effect.get("type") == "heal_hp":
                player.heal(hp=effect.get("value", 0))
                ui.print_line(f"使用了 {item['name']}，恢复 {effect.get('value', 0)} 点气血。")
            elif effect.get("type") == "heal_mp":
                player.heal(mp=effect.get("value", 0))
                ui.print_line(f"使用了 {item['name']}，恢复 {effect.get('value', 0)} 点内力。")
            player.remove_item(item_id, 1)

        # NPC turn
        if npc_hp > 0:
            damage, crit = calculate_damage(npc_attack, player.total_defense(), npc_speed)
            player.hp -= damage
            msg = f"{npc_name} 攻击你，造成 {damage} 点伤害"
            if crit:
                msg += "（暴击！）"
            ui.print_line(msg)

        ui.print_line()

    ui.pause()

    if not player.is_alive():
        ui.print_line("\n你重伤倒地……")
        ui.print_line("一位路过的侠客把你抬回了镇上。")
        player.hp = 1
        player.location_id = "taiping_town_center"
        ui.pause()
        return False

    if npc_hp <= 0:
        ui.print_line(f"\n你击败了 {npc_name}！")
        exp_gain = npc_data.get("level", 1) * 20 + 10
        gold_gain = 0
        player.gain_exp(exp_gain)
        ui.print_line(f"获得 {exp_gain} 点经验。")

        # Loot
        loot_list = npc_data.get("loot", [])
        items = load_data()["items"]
        for loot in loot_list:
            item_id = loot["item_id"]
            chance = loot.get("chance", 1.0)
            quantity = loot.get("quantity", 1)
            if random.random() < chance:
                player.add_item(item_id, quantity)
                item_name = items.get(item_id, {}).get("name", item_id)
                ui.print_line(f"获得：{item_name} x{quantity}")
                if items.get(item_id, {}).get("type") == "currency":
                    gold_gain += quantity

        if gold_gain > 0:
            ui.print_line(f"获得 {gold_gain} 文铜钱。")

        # Morality choice
        if ui.confirm(f"{npc_name} 倒在地上，要杀死他吗？"):
            player.dead_npcs.add(npc["id"])
            player.morality -= 5
            ui.print_line(f"你杀死了 {npc_name}，善恶值下降。")
        else:
            player.morality += 1
            ui.print_line(f"你放过了 {npc_name}，善恶值上升。")

        ui.pause()
        return True

    return False

File: src/combat.js
/** Turn-based combat system. */

import { randomInt, chance } from './utils.js';

export class Combat {
    constructor(player, npc, world) {
        this.player = player;
        this.world = world;
        this.npc = { ...npc };
        const combat = npc.combat || {};
        this.enemy = {
            id: npc.id,
            name: npc.name,
            hp: combat.hp || 50,
            maxHp: combat.hp || 50,
            mp: combat.mp || 0,
            maxMp: combat.mp || 0,
            attack: combat.attack || 10,
            defense: combat.defense || 5,
            speed: combat.speed || 10,
            level: combat.level || 1,
            loot: combat.loot || [],
        };
        this.options = ['攻击', '武功', '物品', '逃跑'];
        this.selectedIndex = 0;
        this.finished = false;
        this.won = false;
        this.log = [`${this.enemy.name} 拦住了你的去路！`];
    }

    handleInput(input) {
        if (this.finished) return;

        if (input.isPressed('up')) {
            this.selectedIndex = (this.selectedIndex - 1 + this.options.length) % this.options.length;
        } else if (input.isPressed('down')) {
            this.selectedIndex = (this.selectedIndex + 1) % this.options.length;
        } else if (input.isPressed('a')) {
            this._executeAction();
        } else if (input.isPressed('b')) {
            this._tryFlee();
        }
    }

    _executeAction() {
        const action = this.options[this.selectedIndex];

        if (action === '攻击') {
            this._playerAttack(false);
        } else if (action === '武功') {
            this._playerAttack(true);
        } else if (action === '物品') {
            this._useItem();
            return;
        } else if (action === '逃跑') {
            this._tryFlee();
            return;
        }

        if (!this.finished && this.enemy.hp > 0) {
            this._enemyTurn();
        }
    }

    _playerAttack(useSkill) {
        let attack = this.player.totalAttack(this.world.items);
        if (useSkill) {
            // Simplified: add skill levels to attack
            let skillBonus = 0;
            for (const [skillId, level] of Object.entries(this.player.skills)) {
                const skill = this.world.getSkill(skillId);
                if (skill && ['unarmed', 'sword', 'blade'].includes(skill.type)) {
                    skillBonus += level * 2;
                }
            }
            if (skillBonus === 0) {
                this.log.push('你还没有学会攻击类武功。');
                return;
            }
            attack += skillBonus;
        }

        const damage = this._calculateDamage(attack, this.enemy.defense, this.player.speed);
        this.enemy.hp -= damage;
        this.log.push(`你攻击 ${this.enemy.name}，造成 ${damage} 点伤害。`);

        if (this.enemy.hp <= 0) {
            this._win();
        }
    }

    _useItem() {
        const consumables = Object.entries(this.player.inventory)
            .filter(([id, qty]) => qty > 0 && this.world.items[id]?.type === 'consumable');

        if (consumables.length === 0) {
            this.log.push('没有可使用的物品。');
            return;
        }

        // Auto-use first healing item for simplicity in this version
        const [itemId, qty] = consumables[0];
        const item = this.world.items[itemId];
        const effect = item.effect || {};

        if (effect.type === 'heal_hp') {
            this.player.heal(effect.value || 0, 0);
            this.log.push(`使用 ${item.name}，恢复 ${effect.value} 点气血。`);
        } else if (effect.type === 'heal_mp') {
            this.player.heal(0, effect.value || 0);
            this.log.push(`使用 ${item.name}，恢复 ${effect.value} 点内力。`);
        }
        this.player.removeItem(itemId, 1);
        this._enemyTurn();
    }

    _tryFlee() {
        const fleeChance = 0.3 + (this.player.speed - this.enemy.speed) * 0.02;
        if (chance(fleeChance)) {
            this.log.push('你趁机逃走了！');
            this.finished = true;
            this.won = false;
        } else {
            this.log.push('逃跑失败！');
            this._enemyTurn();
        }
    }

    _enemyTurn() {
        const damage = this._calculateDamage(this.enemy.attack, this.player.totalDefense(this.world.items), this.enemy.speed);
        this.player.hp -= damage;
        this.log.push(`${this.enemy.name} 攻击你，造成 ${damage} 点伤害。`);

        if (this.player.hp <= 0) {
            this.player.hp = 1;
            this.log.push('你重伤倒地，被抬回了镇上。');
            this.player.locationId = 'taiping_town_center';
            this.player.x = 8;
            this.player.y = 8;
            this.finished = true;
            this.won = false;
        }
    }

    _calculateDamage(attack, defense, speed) {
        const base = Math.max(1, attack - defense * 0.5);
        const variance = 0.9 + Math.random() * 0.2;
        const critChance = speed * 0.005;
        const crit = Math.random() < critChance ? 2 : 1;
        return Math.max(1, Math.floor(base * variance * crit));
    }

    _win() {
        this.finished = true;
        this.won = true;
        const expGain = this.enemy.level * 20 + 10;
        this.player.gainExp(expGain);
        this.player.potential += this.enemy.level * 5;
        this.log.push(`击败 ${this.enemy.name}，获得 ${expGain} 经验。`);

        // Loot
        for (const loot of this.enemy.loot) {
            if (chance(loot.chance)) {
                const qty = loot.quantity || 1;
                this.player.addItem(loot.item_id, qty);
                const itemName = this.world.items[loot.item_id]?.name || loot.item_id;
                this.log.push(`获得：${itemName} x${qty}`);
            }
        }
    }

    chooseKillOrSpare(callback) {
        // Simplified: auto-spare for now; can be extended with UI choice
        this.player.morality += 1;
        this.log.push(`你放过了 ${this.enemy.name}。`);
        callback();
    }
}
